### ISSUE TITLE
Undo function

### ISSUE THREAD

**Issue by Junonia** [qupath/qupath#75]
First thanks for this great application.  I stumbled on QuPath several days ago and tried out. It is really useful for editing the boundary of the cells.  I like the brush tool a lot. The scripting function is easy to learn and powerful. There are several enhancements that will even make this work better for me.

1. Undo function.  Sometimes I deleted an object by mistake, and it will be great if I can get it back.
2. A select tool by drawing a square like many other editing tools.  This will make select multiple objects easy.
3. A point tool allow using points to mark an object.  Some times, we would like just mark the key point of an object.  It is possible to use a polygon but a point tool will be nice.

Thanks again for this nice piece of work.

**Comment by Svidro** [comment]
Ah, Peter just pointed this out to me, but you can try using Shift+E to perform a quick Undo, though I think it only goes back one step.  It seems to work with the point tool as well, which is the three small circles in a triangle on the right-most side of the drawing tools. 

Selection can be an issue, so far I have had to find work arounds with scripting in order to select multiple objects that are not nicely next to each other in the hierarchy.

**Comment by petebankhead** [comment]
Hi there,

> 1. Undo function.  Sometimes I deleted an object by mistake, and it will be great if I can get it back.

An undo option would be useful; unfortunately, there are so many things that you might do with QuPath - especially if you include scripting - that are not easily 'undo-able' without badly impacting performance and increasing the potential for bugs.  For example, there are lots of ways you might change the [object hierarchy](https://github.com/qupath/qupath/wiki/Object-hierarchies) and keeping track of them would require a lot of additional overhead.  So for now... I'm afraid there's no undo.

Nevertheless, there are a few things that can help:

* If you draw something by accident, press *backspace* to delete it
* If you delete something by accident, as @Svidro says, press ```Shift+E```.  This corresponds to the *Restore Selection* command in ImageJ, and has the same shortcut, so as to help if you happen to know ImageJ already.  (However, it only ever remembers the most recent object that it saw... so if you select any other object in the meantime, then it cannot restore the one that was deleted.)
* If you want to reduce the chances of accidentally editing an annotation, right-click and choose *Annotations &rarr; Lock*.  Note that the annotation can still be deleted - but not moved or otherwise edited.
* Regularly save (```Ctrl + S```) and then use *File &rarr; Revert* to go back (```Ctrl + R```).

The last one is probably the approach I use most.

> 2. A select tool by drawing a square like many other editing tools.  This will make select multiple objects easy.

I've also wanted this occasionally, but not often enough to do anything about it yet.  However, there are a few options for selecting:

* Click on each object in either the *Annotations* or *Hierarchy* tab (with ```Ctrl``` or ```Shift``` pressed if needed)
* If you have one object selected, and the *Move* tool is active, you can select other objects by clicking them in the image with the ```Alt``` key pressed
* You can use a script, such as the one below

```groovy
guiscript=true
// Get the current selected object & hierarchy
selected = getSelectedObject()
hierarchy = getCurrentHierarchy()
// Get all the objects inside the current selection
objectsToSelect = hierarchy.getDescendantObjects(selected, null, null)
if (objectsToSelect != null) {
    // Remove the current selected object
    hierarchy.removeObject(selected, true)
    // Update the selection
    hierarchy.getSelectionModel().selectObjects(objectsToSelect)
}
```

This should enable you to draw around objects using any of the drawing tools, and then run the script to select everything inside.  As it is written, the annotation that you draw will then be deleted.

This has the advantage of making it possible to select objects by drawing any arbitrary shape, without needing to add an extra tool to the toolbar.

If this were to become a standard command, with a shortcut, would it address the need?  Or do you think a separate selection tool would still be preferable?

> 3. A point tool allow using points to mark an object.  Some times, we would like just mark the key point of an object.  It is possible to use a polygon but a point tool will be nice.

There is indeed a *Point tool*, described [here](https://github.com/qupath/qupath/wiki/Counting-cells#manual-cell-counting).

**Comment by Svidro** [comment]
Another useful script to snag!  Is there a getDescendantAnnotations of any sort or would the best option be to run another loop to check all descendant objects for being an annotation, if I just wanted all of my annotations but not the SLICs or cells.

**Comment by petebankhead** [comment]
No separate method, but this should do the trick:
```groovy
objectsToSelect = hierarchy.getDescendantObjects(selected, null, qupath.lib.objects.PathAnnotationObject)
```

**Comment by Junonia** [comment]
Thanks for the replies,  scripts and pointed out the point tool.  Is there a easy way to delete individual points in the group.

**Comment by petebankhead** [comment]
With the *Point* tool selected, click on an existing point while holding down the ```Alt``` key, as described under [Tips for more effective counting](https://github.com/qupath/qupath/wiki/Counting-cells#tips-for-more-effective-counting).

**Comment by petebankhead** [comment]
Some undo support is now available in v0.2.0-m1 (and beyond), and explained [on my blog](https://petebankhead.github.io/qupath/2018/03/19/qupath-updates.html#new-limited-support-for-undo).
